Compile Doc Tests

killercup opened the pull request:
I was looking through the code and noticed there ere no tests and all doc tests were disabled. So, I fixed the doc tests and enabled them. :smile: 

Fixes #50

japaric commented on src/lib.rs:
````diff
@@ -59,12 +59,14 @@ impl Bencher {
     ///
-    /// ``` ignore
-    /// start = Instant::new();
+    /// ```rust,no_run
````
What does the `rust` do here (and in all the other occurrences below)? I haven't used it before.

killercup replied:
It's just Markdown's syntax to set the language used in the code block (just like `text` below). I like to add it to make things more explicit. The `ignore` or `no_run` flags only work when `rust` is set of course – rustdoc assumes `rust` by default when no language is set.

japaric commented on src/lib.rs:
````diff
@@ -59,12 +59,14 @@ impl Bencher {
-    /// start = Instant::new();
+    /// ```rust,no_run
+    /// # fn routine() {}
+    /// let iters = 4_000_000;
````
Hide this variable with `#`. The number of iterations is chosen by the `Bencher` at runtime, this gives the impression that `iters` is fixed.

japaric commented on src/lib.rs:
````diff
@@ -59,12 +59,14 @@ impl Bencher {
+    /// ```rust,no_run
+    /// # fn routine() {}
+    /// let iters = 4_000_000;
+    /// let start = std::time::Instant::now();
````
I think it would be better to import `std::time::Instant` to avoid the full path here.

killercup replied:
I'm gonna change this to (hidden) imports – that's what you meant, right?

japaric commented on src/lib.rs:
```diff
@@ -59,12 +59,14 @@ impl Bencher {
     ///     routine();
     /// }
-    /// elapsed = start.elapsed();
+    /// println!("{:?}", start.elapsed());
```
revert this back to the `elapsed` binding because the "Timing model" section below refers to it.

japaric commented on src/lib.rs:
````diff
@@ -127,18 +129,21 @@ impl Bencher {
+    /// ```rust,no_run
+    /// # fn setup() {}
+    /// # fn routine(input: ()) {}
+    /// let iters = 4_000_000;
````
hide this for the same reason stated above.

japaric commented on src/lib.rs:
```diff
@@ -127,18 +129,21 @@ impl Bencher {
+    /// # fn setup() {}
+    /// # fn routine(input: ()) {}
+    /// let iters = 4_000_000;
+    /// let elapsed = std::time::Duration::new(0, 0);
```
as above, importing `Duration` would remove the need to use a full path here.

japaric replied:
hmm, shouldn't `elapsed` need to be mutable because of the re-assignment below?

japaric commented on src/lib.rs:
```diff
@@ -127,18 +129,21 @@ impl Bencher {
     ///     let input = setup();
     ///
-    ///     let start = Instant::now();
+    ///     let start = std::time::Instant::now();
```
ditto, import - full path.

japaric commented on src/lib.rs:
```diff
@@ -127,18 +129,21 @@ impl Bencher {
     ///     let elapsed_in_iter = start.elapsed();
     ///
-    ///     mem::drop(output);
+    ///     std::mem::drop(output);
```
ditto, import `std::mem` - use `mem::drop`.

japaric commented on src/lib.rs:
````diff
@@ -177,16 +182,18 @@ impl Bencher {
-    /// ``` ignore
-    /// outputs = Vec::with_capacity(iters);
+    /// ```rust,no_run
+    /// let iters = 4_000_000;
````
ditto. hide.

japaric commented on src/lib.rs:
```diff
@@ -177,16 +182,18 @@ impl Bencher {
+    /// let outputs = Vec::with_capacity(iters);
     ///
-    /// start = Instant::now();
+    /// let start = std::time::Instant::now();
```
ditto. import, full path.

japaric commented on src/lib.rs:
```diff
@@ -177,16 +182,18 @@ impl Bencher {
     ///     outputs.push(routine());
     /// }
-    /// elapsed = start.elapsed();
+    /// println!("{:?}", start.elapsed());
```
ditto, revert to `elapsed` assignment.

japaric commented on src/lib.rs:
```diff
@@ -177,16 +182,18 @@ impl Bencher {
+    /// println!("{:?}", start.elapsed());
     ///
-    /// mem::drop(outputs);
+    /// std::mem::drop(outputs);
```
ditto. import `std::mem`.

japaric commented on src/lib.rs:
````diff
@@ -221,14 +228,17 @@ impl Bencher {
+    /// ```rust,no_run
+    /// # fn setup() {}
+    /// # fn routine(input: ()) {}
+    /// let iters = 4_000_000;
````
ditto. hide.

japaric commented on src/lib.rs:
```diff
@@ -221,14 +228,17 @@ impl Bencher {
+    /// let inputs: Vec<()> = (0..iters).map(|_| setup()).collect();
     ///
-    /// start = Instant::now();
+    /// let start = std::time::Instant::now();
```
ditto. import.

japaric commented on src/lib.rs:
```diff
@@ -221,14 +228,17 @@ impl Bencher {
     ///     routine(input);
     /// }
-    /// elapsed = start.elapsed();
+    /// println!("{:?}", start.elapsed());
```
ditto. `elapsed` assignment.

japaric commented:
Thanks for the PR! The "timing loops" were more like pseudo-code in my mind, but I think it's a great idea to test them :+1:. I left a bunch of comments :smiling_imp:.

killercup commented:
I think I addressed all your comments :smiley:

I originally wanted to use full paths everywhere so people knew this was from `std`, by the way. I don't think that's really necessary, though, so I changed it to (hidden) imports.

japaric commented:
Thanks! :+1:
